RTSP Client basic fucntionality

Hajto opened the pull request:
Implements basic RTSP client with pipeable TCP Socket. (pipeable meaning you can pipe many requests through single open connection)

## Testing

Tests are implemented on 3 layers: unit, integration and external integration.
By default external tests are disabled. They are meant to be used only manually in development to make absolutely sure the client works with real RTSP server. Integration tests were implemented by mocking external resource with responses aggregated from external tests.

## Possibilities of improvment in future.
If both Session and Transport would be modified to use request queueing under the hood, then requests could be resolved asynchronously (https://www.annkissam.com/elixir/alembic/posts/2018/07/30/genserver-reply.html).

But that is not needed right now and current implementation should be sufficient.

## Suggested review stages

This PR is rather bulky so I suggest reviewing it in following stages basing on the modules software architecture.

- [x] Transport (Transport, PipeableTCPSocket)
- [x] Request and Response formatting
- [x] Session lifecycle and request execution
- [ ] RTSP facade and integration

Hajto committed:
Basic rtsp functionality

Hajto committed:
Added pipeable tcp socket and sdp parsing.

Hajto committed:
Changed RTSP interface, added transport supervision

Hajto committed:
Added support for User Agent header, minor typespecs and bugfixes, added integration texts.

Hajto committed:
Closing Session process lifecycle.

Hajto committed:
Added RTSP facade. Added starting and stoping rtsp containers.

Hajto committed:
Removed Bunch where it wasn't needed added documentation.

Hajto committed:
Basic documentation added.

Hajto committed:
Added documenation.
Refactored request to avoid `to_string` override.

Hajto committed:
Added CircleCI

Hajto committed:
Added documentation and specs.
Credo based alias cleanup.

Hajto committed:
Fixed RTSP specs.

Hajto committed:
Remove redundant types.

Hajto committed:
Removed queueing requests since Session resolves them in blocking manner.

Hajto committed:
Slight style refacor.

Hajto commented:
@bblaszkow06 @mat-hek I have no idea what I could write about in Response and Request moduledocs, so I would be grateful for any suggestions prior actual review.

Hajto committed:
Bumped bunch to 1.0

Hajto committed:
Overhauled documentation.

bblaszkow06 commented on lib/rtsp.ex:
```diff
@@ -0,0 +1,81 @@
+defmodule Membrane.Protocol.RTSP do
+  @moduledoc """
+  This module provides a functionality to open a sessi on, execute a RTSP requests
```
remove space from `sessi on`

bblaszkow06 commented on README.md:
```diff
@@ -1,11 +1,58 @@
+
+To implement custom request execution logic you must implement
+`Membrane.Protocol.RTSP.Transport` behaviour. Then you can pass
+name of your transport module to `Membrane.Protocol.RTSP.start/3` as second argument.
```
the name ... as a second argument

bblaszkow06 commented on lib/rtsp/transport.ex:
```diff
@@ -0,0 +1,28 @@
+  @doc """
+  Invoked by session process when executing requests.
+  """
+  @callback execute(request :: binary(), transport_ref, options :: [tuple()]) ::
```
Quite unusual typespec for options, why not keyword?

bblaszkow06 commented on lib/rtsp/transport/pipeable_tcp_socket.ex:
```diff
@@ -0,0 +1,84 @@
+  end
+
+  @spec open(URI.t()) :: {:error, atom()} | {:ok, :gen_tcp.socket()}
+  defp open(%URI{host: host, port: port}) do
```
Move this function to the bottom of module

bblaszkow06 commented on lib/rtsp/transport/pipeable_tcp_socket.ex:
```diff
@@ -0,0 +1,84 @@
+defmodule Membrane.Protocol.RTSP.Transport.PipeableTCPSocket do
```
I'm not a big fan of this name, what do you think of `PersistentTCPSocket`? Or even simply `TCPSocket` since it's an implementation detail that the socket is kept alive

Hajto replied:
@bblaszkow06 `Pipeable` in the name refers to the ability to pipe as many requests as you want into a single open connection and not reestablishing it every single time.

I don't think `Persistent` is a good name for it either it could possibly suggest that it is a mechanism similar to persistent mq.

bblaszkow06 commented on lib/rtsp/transport/pipeable_tcp_socket.ex:
```diff
@@ -0,0 +1,84 @@
+
+  @spec open(URI.t()) :: {:error, atom()} | {:ok, :gen_tcp.socket()}
+  defp open(%URI{host: host, port: port}) do
+    mockable(:gen_tcp).connect(to_charlist(host), port, [:binary, {:active, true}])
```
You can leverage Elixir syntax here (`binary: {:active, true}`)

bblaszkow06 commented on lib/rtsp/transport/pipeable_tcp_socket.ex:
```diff
@@ -0,0 +1,84 @@
+  @impl true
+  def terminate(reason, state)
+  def terminate(_, %State{connection: nil}), do: :ok
+  def terminate(_, %State{connection: connection}), do: mockable(:gen_tcp).close(connection)
```
From https://hexdocs.pm/elixir/GenServer.html#c:terminate/2 :

>There is no cleanup needed when the GenServer controls a port (e.g. :gen_tcp.socket) or File.io_device/0, >because these will be closed on receiving a GenServer's exit signal and do not need to be closed manually in >terminate/2.

So this makes not sense (+ it is not called on shutdown since you don't trap exits)

bblaszkow06 commented on lib/rtsp/request.ex:
```diff
@@ -0,0 +1,86 @@
+    list
+    |> Enum.map(fn elem -> header_to_string(elem) end)
+    |> Enum.join("\r\n")
+    ~> ("\r\n" <> &1)
```
A non-bunch way would be 
`|> String.replace_prefix("", "\r\n")`

bblaszkow06 commented on lib/rtsp/request.ex:
```diff
@@ -0,0 +1,86 @@
+  defp apply_path(url, %__MODULE__{path: nil}), do: url
+
+  defp apply_path(url, %__MODULE__{path: path}),
+    do: Path.join(url, path)
```
Shouldn't that be https://hexdocs.pm/elixir/URI.html#merge/2? (if used before `to_string`)

Hajto replied:
I will explain using examples why that's not ideal solution to use merge.
We have:
```elixir
uri = %URI{
  authority: "domain.net:554",
  fragment: nil,
  host: "domain.net",
  path: "/path:file.mov",
  port: 554,
  query: nil,
  scheme: "rtsp",
  userinfo: nil
}
```
after merge:
```elixir
iex> URI.merge(uri, "trackID=2")
%URI{
  authority: "domain.net:554",
  fragment: nil,
  host: "domain.net",
  path: "/trackID=2",
  port: 554,
  query: nil,
  scheme: "rtsp",
  userinfo: nil
}
```

Hajto replied:
So merge kinda replaces path field in URI struct.

Hajto replied:
and we want to merge path :| (because one path is for the resource that session operates on and path on request is surfix to that)

Hajto replied:
Additional path was added to enable executing request over control urls described here: https://tools.ietf.org/html/rfc2326#appendix-C.3

Hajto replied:
Handling absolute controls is a feature for separate PR I think, because most probably we won't need it (V IP cameras didn't seem to use them)

bblaszkow06 commented on lib/rtsp/request.ex:
```diff
@@ -0,0 +1,86 @@
+
+  defp render_headers(list) do
+    list
+    |> Enum.map(fn elem -> header_to_string(elem) end)
```
You can use https://hexdocs.pm/elixir/Enum.html#map_join/3 to spare one iteration

bblaszkow06 commented on lib/rtsp/request.ex:
```diff
@@ -0,0 +1,86 @@
+    ~> ("\r\n" <> &1)
+  end
+
+  defp header_to_string({header, value}), do: header <> ": " <> to_string(value)
```
Typespec says value has to be binary, so to_string should not be needed

bblaszkow06 commented on lib/rtsp/response.ex:
```diff
@@ -0,0 +1,114 @@
+    [line, rest] = split_next_chunk(binary)
+
+    case String.split(line, ":", parts: 2) do
+      [name, " " <> value | []] -> parse_headers({response, rest}, [{name, value} | acc])
```
`| []` should be removed

bblaszkow06 commented on lib/rtsp/response.ex:
```diff
@@ -0,0 +1,114 @@
+    end
+  end
+
+  defp split_next_chunk(response), do: String.split(response, "\r\n", parts: 2)
```
Wouldn't it be simpler to:
1. split response by first `\r\n\r\n`
1. split headers by `\r\n` and do Enum.map or Bunch.Enum.try_map without recursion, accumulator and reversing
1. parse body

or am I missing some corner case?

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -0,0 +1,124 @@
+    transport_ref = Transport.transport_name(executor)
+
+    request
+    |> Request.with_header("CSeq", cseq)
```
I like this "builder" API :+1:

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -0,0 +1,124 @@
+            transport: module(),
+            cseq: non_neg_integer(),
+            uri: URI.t(),
+            transport_executor: binary(),
```
The naming here is quite inconsistent. This is sometimes `name`, `ref`, here it is called executor for some reason. Please unify that and use only one name in the whole element for each:
- key used to store transport process in registry (`transport_key`/`key`)
- tuple used in send ({`:via, {Registry, blahblahblah...`) - i would opt for reference as here https://hexdocs.pm/elixir/GenServer.html#t:server/0

bblaszkow06 replied:
Plus, I see that the transport module and key are often passed together. Maybe that deserves a `Transport` struct? You could then reference that nicely as `transport.module` and `transport.key`

bblaszkow06 commented on lib/rtsp/transport.ex:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  @spec transport_name(binary()) :: transport_ref
+  def transport_name(ref), do: {:via, Registry, {TransportRegistry, ref}}
```
Naming inconsistency, see comment on Session

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -0,0 +1,124 @@
+  and `User-Agent` header. If the URI contains credentials they will also
+  be added unless `Authorization` header is present in request.
+  """
+  @spec execute(pid(), Request.t(), non_neg_integer()) :: {:ok, Response.t()} | {:error, atom()}
```
Requests (in this context) are rather sent than executed. Eventually, you can use request as a verb https://hexdocs.pm/httpotion/HTTPotion.html#request/3

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -0,0 +1,124 @@
+    end
+  end
+
+  defp perform_execution(request, %State{uri: uri, execution_options: options} = state) do
```
Who is killed by this function and how? Guillotine? Hanging? Electric chair?

Hajto replied:
It was more like perform performance, but I get the idea.

Hajto committed:
Added note about external tests to Readme.md

Hajto committed:
Added SoftwareMansion footer.

bblaszkow06 reviewed (changes requested):
To be continued..
- [x] Supervision
- [x] tests

Hajto committed:
Fixed docs and typespecs.
Fixed bug in session and corresponding test where non binary value could be passed.
Moved code around in TCP Socket.

Hajto commented on test/support/mockery_helper.ex:
```diff
@@ -0,0 +1,31 @@
+    |> ExUnit.Assertions.assert("""
+    #{Utils.print_mod(mod)}.#{fun} was not called with arguments matching criteria.
+    """)
+  end
```
This function can be potentially submitted to mockery as PR.

Hajto committed:
Transport ref and module pluming refactor.

Hajto committed:
Moved from recursive response parsing to more iterative approach.

Hajto committed:
Renamed `PipeableTCPSocket` to `TCPSocket`

Hajto committed:
Renamed execute in `Session` and do_request in `RTSP`

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -29,18 +28,17 @@ defmodule Membrane.Protocol.RTSP.Session do
-    * transport - module that is responsible for executing request
-    * transport_executor - a reference (`Registry` key) that will be used
-    when executing request
+    * transport - struct representing module that will be used when executing
```
This description doesn't make much sense here. Such doc belongs rather to Transport module

Hajto replied:
@bblaszkow06 So it should be replaced with reference to `Transport` module?

bblaszkow06 replied:
Seems like a good idea

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -81,16 +78,15 @@ defmodule Membrane.Protocol.RTSP.Session do
     |> apply_credentials(uri)
     |> Request.stringify(uri)
-    |> transport.execute(transport_ref, options)
+    |> transport.module.execute(transport.key, options)
```
Maybe there should be a wrapper in Transport that would do this?

bblaszkow06 commented on README.md:
```diff
@@ -15,7 +64,18 @@ def deps do
+
+## Copyright and License
+
+Copyright 2018, [Software Mansion](https://swmansion.com/?utm_source=git&utm_medium=readme&utm_campaign=membrane)
```
Minor nitpicking - this should be 2019

bblaszkow06 commented on lib/rtsp/response.ex:
```diff
@@ -0,0 +1,106 @@
+  """
+  @spec parse(binary()) :: {:ok, t()} | {:error, :invalid_start_line | :malformed_header}
+  def parse(response) do
+    [headers, body] = String.split(response, "\r\n\r\n", parts: 2)
```
As you mentioned f2f, you need to support `\n\n` as well (maybe regex will do?)

bblaszkow06 commented on lib/rtsp/response.ex:
```diff
@@ -0,0 +1,106 @@
+  @doc """
+  Parses RTSP response.
+
+  Requires a raw response to be `\\r\\n` delimited. If the body is present
```
Remember to fix the docs as well

bblaszkow06 commented on lib/rtsp/response.ex:
```diff
@@ -0,0 +1,106 @@
+
+  defp parse_headers(headers) do
+    headers
+    |> String.split("\r\n")
```
\+ `\n`

bblaszkow06 commented on lib/rtsp/session/couple_supervisor.ex:
```diff
@@ -0,0 +1,50 @@
+          |> :os.system_time()
+          |> to_string()
+
+        transport = Transport.new(transport, time <> raw_url)
```
If you need a unique identifier, wouldn't it be simpler to use `make_ref`?

bblaszkow06 commented on lib/rtsp/session/supervisor.ex:
```diff
@@ -0,0 +1,26 @@
+  end
+
+  @spec terminate_child(any()) :: :ok | {:error, :not_found | :simple_one_for_one}
+  def terminate_child(pid), do: Supervisor.terminate_child(__MODULE__, pid)
```
I'm pretty sure you wanted to call `DynamicSupervisor.terminate_child/2` and that would change the spec as well

bblaszkow06 commented on lib/rtsp/session/couple_supervisor.ex:
```diff
@@ -0,0 +1,50 @@
+defmodule Membrane.Protocol.RTSP.Session.CoupleSupervisor do
```
I'm not a big fan of this name. I would suggest renaming current `Session.Supervisor` to `RTSP.Supervisor` and this one to `Session.Supervisor`

bblaszkow06 commented on lib/rtsp.ex:
```diff
@@ -0,0 +1,81 @@
+
+  @spec start(binary(), module(), Keyword.t()) :: :ignore | {:error, atom()} | {:ok, t()}
+  def start(url, transport \\ TCPSocket, options \\ []) do
+    with {:ok, supervisor} <- Session.Supervisor.start_child(transport, url, options) do
```
the `supervisor` name is quite ambiguous. Rename it to match that module name

bblaszkow06 commented on test/rtsp/session/session_integration_test.exs:
```diff
@@ -0,0 +1,101 @@
+      Session: 443551157;timeout=60
+
+      v=0
+      o=- 443551157 443551157 IN IP4 184.72.239.149\r\ns=BigBuckBunny_115k.mov
```
why is there `\r\n`?

bblaszkow06 commented on test/support/mockery_helper.ex:
```diff
@@ -0,0 +1,31 @@
+defmodule Membrane.Support.MockeryHelper do
```
This module is an abomination and should be removed.
1. You are using undocumented, private API!
1. It is not needed, you can add assertions in mock:
```elixir
mock Foo, [bar: 1], fn some_binary ->
  assert String.contains?(some_binary, "Some magic criteria")
end
```

Hajto committed:
Update year in Readme and added License

Hajto committed:
Added support for all line endings.

Hajto committed:
Replaced uid based on time to make_ref

Hajto committed:
Replaced Mockery Helper with asserts in mock definitions.

Hajto committed:
Supervisor rename.

Hajto committed:
First step of major module restructuring.

Hajto committed:
Session is now container and process handling lifecycles is called manager.

Hajto committed:
Session now contains information which RTSP had and enables to send any request.

Session also contains serves as a container for Session and Transport combination.

Hajto committed:
Refactor cleanup.

Hajto committed:
Few minor naming changes

Hajto committed:
Fixed missing alias for TCPSocket

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -1,124 +1,79 @@
-      {"Authorization", _} ->
-        request
+  @doc """
+  Starts and links process that Session Manager and companion Transport process.
```
Some word is missing here

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -1,124 +1,79 @@
+  """
+  @spec start_container(module(), binary(), Keyword.t()) ::
+          Supervisor.on_start() | {:error, :invalid_url}
+  def start_container(transport, raw_url, options) do
```
Since this function calls `Supervisor.start_link` I would follow the convention and called `start_link`

Hajto replied:
resolved by extracting container

bblaszkow06 commented on lib/rtsp/session.ex:
```diff
@@ -1,124 +1,79 @@
-
-    {:ok, state}
+  @spec start(binary(), module(), Keyword.t()) :: :ignore | {:error, atom()} | {:ok, t()}
+  def start(url, transport \\ TCPSocket, options \\ []) do
```
This fun. name is a bit confusing when paired with `start_container`. I would call it either `new`, `spawn` or `create`

bblaszkow06 commented on lib/rtsp/session/manager.ex:
```diff
@@ -0,0 +1,119 @@
+  Starts and links session process.
+
+  Sets following properties of Session:
+    * transport - information for executing request over the network for
```
I would end the sentence after network

bblaszkow06 commented on lib/rtsp/transport.ex:
```diff
@@ -1,10 +1,26 @@
   use Bunch
+  defstruct [:module, :key]
 
   @type transport_ref :: {:via, Registry, {TransportRegistry, binary()}}
```
Now the type for ref is reference, not binary

Hajto committed:
Added timeout param to connect to prevent neverending freezes.

Hajto committed:
Documentation cleanup.

bblaszkow06 reviewed (approved):
Just a tiny detail - update mix.lock as it is broken (references commit in SDP that doesn't exist)

mat-hek commented on lib/rtsp/application.ex:
```diff
@@ -0,0 +1,21 @@
+        id: TransportRegistry,
+        start: {Registry, :start_link, [:unique, TransportRegistry]}
+      },
+      %{
```
I suppose you can pass just `RTSP.Supervisor` instead of this map

mat-hek commented on lib/rtsp/request.ex:
```diff
@@ -0,0 +1,83 @@
+
+  defp render_headers(list) do
+    list
+    |> Enum.map_join("\r\n", fn elem -> header_to_string(elem) end)
```
`&header_to_string/1`

mat-hek commented on lib/rtsp/response.ex:
```diff
@@ -0,0 +1,108 @@
+  @type t :: %__MODULE__{
+          status: non_neg_integer(),
+          headers: Membrane.Protocol.RTSP.headers(),
+          body: any()
```
shouldnt'it be `SDP.Session.t | String.t`?

mat-hek commented on lib/rtsp/session/manager.ex:
```diff
@@ -0,0 +1,120 @@
+  end
+
+  defp execute(request, %State{uri: uri, execution_options: options} = state) do
+    %State{cseq: cseq, transport: transport} = state
```
unpack all the values here instead of header

mat-hek commented on README.md:
````diff
@@ -1,11 +1,60 @@
+```elixir
+alias Membrane.Protocol.RTSP
+
+{:ok, session} = RTSP.start("rtsp://domain.name:port/path")
````
`RTSP.Session.start`

mat-hek commented on lib/rtsp/application.ex:
```diff
@@ -0,0 +1,21 @@
+      }
+    ]
+
+    Supervisor.start_link(children, strategy: :one_for_one)
```
Add a name

Hajto replied:
@mat-hek Why?

mat-hek replied:
to make recognizable in observer

mat-hek commented on lib/rtsp/session/manager.ex:
```diff
@@ -0,0 +1,120 @@
+  end
+
+  # Some responses does not have to return Session ID
+  # If it does return one it needs to match one stored in state
```
fix english

mat-hek commented on .circleci/config.yml:
```diff
@@ -0,0 +1,14 @@
+      - run: mix local.hex --force
+      - run: mix deps.get
+      - run: mix format --check-formatted
+      - run: mix test
```
I think external tests can be included here

Hajto replied:
I don't want to load someone's server in recurrent automatic tests.

mat-hek commented on README.md:
```diff
@@ -1,11 +1,60 @@
+
+To implement custom request execution logic you must implement
+`Membrane.Protocol.RTSP.Transport` behaviour. Then you can pass
+the name of your transport module to `Membrane.Protocol.RTSP.start/3` as
```
Session.start

mat-hek commented on README.md:
```diff
@@ -1,11 +1,60 @@
+the name of your transport module to `Membrane.Protocol.RTSP.start/3` as
+a second argument.
+
+`Membrane.Protocol.RTSP.start/3` assumes that transport module also implements
```
Session.start

mat-hek commented on lib/rtsp/session/manager.ex:
```diff
@@ -0,0 +1,120 @@
+defmodule Membrane.Protocol.RTSP.Session.Manager do
+  @moduledoc """
```
The same case as `Session.Container`

Hajto replied:
@mat-hek Technically if someone wanted to use stateless Transport it could be used. The question is then, whether we want to allow it or encourage it. Having ready to go solution it is highly unlikely that scenario will happen.

mat-hek commented on lib/rtsp/session/container.ex:
```diff
@@ -0,0 +1,42 @@
+defmodule Membrane.Protocol.RTSP.Session.Container do
```
Is this module usable by the end user anyhow? Shouldn't it be `@moduledoc false`?

Hajto replied:
It won't be usefull when start_link will be introduced to the session.

mat-hek commented on README.md:
```diff
@@ -15,7 +64,18 @@ def deps do
+Tests that use external RTSP service are disabled by default but they are present
+in the codebase. They are tagged as external and are usually accompanied by
+tests that mimic they behaviour by using predefined responses.
+
```
I would add a chapter describing the architecture of this in two-three sentences

mat-hek commented on lib/rtsp/session.ex:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  @spec request(t(), binary(), RTSP.headers(), binary(), nil | binary()) :: Response.result()
+  def request(session, method, headers \\ [], body \\ "", path \\ nil) do
```
I would move the docs from `Session.Manager.request` to here

mat-hek commented on lib/rtsp/session.ex:
```diff
@@ -0,0 +1,44 @@
+        }
+
+  @doc """
+  Starts Supervised Session.
```
I suppose there is a possibility to elaborate a bit more about this ;)

mat-hek commented on lib/rtsp/session.ex:
```diff
@@ -0,0 +1,44 @@
+  Starts Supervised Session.
+  """
+  @spec start(binary(), module(), Keyword.t()) :: :ignore | {:error, atom()} | {:ok, t()}
+  def start(url, transport \\ Transport.TCPSocket, options \\ []) do
```
Shouldn't there be a `start_link` counterpart?

mat-hek commented on lib/rtsp/session.ex:
```diff
@@ -0,0 +1,44 @@
+  """
+  @spec start(binary(), module(), Keyword.t()) :: :ignore | {:error, atom()} | {:ok, t()}
+  def start(url, transport \\ Transport.TCPSocket, options \\ []) do
+    with {:ok, supervisor} <- RTSPSupervisor.start_child(transport, url, options) do
```
Not sure if regular `start` should start a process under some global supervisor by default. For example, doc for `GenServer.start` says
> Starts a GenServer process without links (outside of a supervision tree).

Hajto committed:
Documentation and typing fixes.

Hajto committed:
Improved documentation, Session now is started under Supervisor that is provided explicitly.

Hajto committed:
Start_link returns value instead of {:ok, value} tuple, because it exits anyway.

Hajto committed:
Applied Credo suggestions and added documentation for Supervisor.

Hajto committed:
Documentation fixes.

Hajto committed:
Fixed SDP git ref in mix.lock

mat-hek commented on README.md:
```diff
@@ -64,13 +59,32 @@ def deps do
+To implement custom request execution logic you must implement
+`Membrane.Protocol.RTSP.Transport` behavior. Then you can pass
+the name of your transport module to `Membrane.Protocol.RTSP.Session.new/4` as
+a second argument.
```
Not as a second argument. Don't provide such details if not necessary

Hajto replied:
There is no unneeded details now.

mat-hek commented on README.md:
```diff
@@ -1,18 +1,22 @@
 # Membrane Protocol RTSP
 
-RTSP client for elixir.
+RTSP client for the elixir.
```
for Elixir

mat-hek reviewed (changes requested):
And go through the comments I previously left and make sure you fixed what was to be fixed, e.g. this https://github.com/membraneframework/membrane-protocol-rtsp/pull/1/files#diff-4e48fd2971100003792baf774c22afb7R13

Hajto committed:
Added elixir_ls to gitignore, edited readme.md

Hajto committed:
Fixed english in Readme and manager.ex
